benchmarks/replay/replay.cpp

```diff
@@ -74,42 +79,128 @@ struct allocation {
  * @tparam MR The type of the `device_memory_resource` to use for allocation
  * replay
  */
-struct replay_benchmark {
+struct replay_benchmark : benchmark::Fixture {
+  MRFactoryFunc factory_;
   std::shared_ptr<rmm::mr::device_memory_resource> mr_{};
   std::vector<std::vector<rmm::detail::event>> const& events_{};
 
+  // Maps a pointer from the event log to an active allocation
+  std::unordered_map<uintptr_t, allocation> allocation_map;
+  std::mutex map_mutex;
+
+  // shared index to play back multithreaded events back in order
+  std::atomic_size_t event_index{0};
+
   /**
    * @brief Construct a `replay_benchmark` from a list of events and
    * set of arguments forwarded to the MR constructor.
    *
+   * @param factory A factory function to create the memory resource
    * @param events The set of allocation events to replay
    * @param args Variable number of arguments forward to the constructor of MR
    */
   replay_benchmark(MRFactoryFunc factory,
                    std::vector<std::vector<rmm::detail::event>> const& events)
-    : mr_{factory()}, events_{events}
+    : factory_{factory}, mr_{}, events_{events}, allocation_map{events.size()}, event_index{0}
+  {
+  }
+
+  /**
+   * @brief Copy construct a replay_benchmark
+   *
+   * Does not copy the mutex or the map
+   */
+  replay_benchmark(replay_benchmark const& other)
```

Comment:
Does this need a copy ctor?

Reply:
Yes because of the mutex. Won't compile without it.

Reply:
I don't understand. Where is `replay_benchmark` being copied that it needs a copy ctor. `std::mutex` is neither copyable or movable, so I understand why the default copy ctor would fail, but I wouldn't think `replay_benchmark` should ever need to be copied. Instead I'd explicitly delete `replay_benchmarks` copy and move ctors.

Reply:
`template <class Lam> benchmark::RegisterBenchmark(const char*, Lam&&)` creates an instance of this class, which invokes the copy ctor of the `Lam` that is passed to it.

```
#ifdef BENCHMARK_HAS_CXX11
template <class Lambda>
class LambdaBenchmark : public Benchmark {
 public:
  virtual void Run(State& st) { lambda_(st); }

 private:
  template <class OLambda>
  LambdaBenchmark(const char* name, OLambda&& lam)
      : Benchmark(name), lambda_(std::forward<OLambda>(lam)) {}

  LambdaBenchmark(LambdaBenchmark const&) = delete;

 private:
  template <class Lam>
  friend Benchmark* ::benchmark::RegisterBenchmark(const char*, Lam&&);

  Lambda lambda_;
};
#endif
```
https://github.com/google/benchmark/blob/4475ff6b8a7a4077d7492b76ef5278a3dc53a2e4/include/benchmark/benchmark.h#L1017-L1036

Reply:
My bad, it invokes the move ctor. I have now replaced this copy ctor with a move ctor and explicitly deleted the copy ctor.